### PR TITLE
Adding ignore bits to narrow compare of received codes

### LINKED ISSUE #292: Configurable number of bits to ignore in code match of rc_switch_raw remote_receiver
**Describe the problem you have/What new integration you would like**

Getting useful results using low quality RF receivers and handheld remotes is impossible if the rc_switch_raw code matching must be exact. In my case the 32 bit received code value is only stable in the top 24 bits, which is plenty to distinguish between the buttons on the remote control. I need to have a way to tell the rc_switch_raw matching process that it should just consider the top 24 bits and discard the bottom 8. This would probably be most useful if it allowed a configurable number of bits (0-31) to be dropped for the comparison.

**Please describe your use case for this integration and alternatives you've tried:**

I've used the existing code at version 1.13.6 but have very limited success, the error rate for any button press is random but typically at least 80% missed button presses, no matter which of the many captured code values I use.

**Additional context**

I've hacked up a test and dropping the lower 8 bits produces >90% success rate on button presses so I'm fairly sure this will work with minimum downside. I'll work on a complete solution and submit a PR.

### PR DESCRIPTION
**Pull request by mtl010957**
## Description:
Low quality receivers and handheld 433MHz RF remotes seem to produce random values for some of (normally the lowest order) bits of the 32 bit code as recognized by the rc_switch_raw configuration of the [Remote Receiver](https://esphome.io/components/remote_receiver.html) platform. Adding ability to define any of the bits to be ignored for the comparison to the expected code for a given input button press makes the recognition much more robust.

**Related issue (if applicable):** fixes [feature request #292](https://github.com/esphome/feature-requests/issues/292)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#279
## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/remote_base/rc_switch_protocol.h [hunk 1]
```diff
@@ -59,6 +59,7 @@ template<typename... Ts> class RCSwitchRawAction : public RemoteTransmitterActio
  public:
   TEMPLATABLE_VALUE(RCSwitchBase, protocol);
   TEMPLATABLE_VALUE(std::string, code);
+  TEMPLATABLE_VALUE(std::string, mask);
```
Comment (OttoWinter) [inline esphome/components/remote_base/rc_switch_protocol.h]: You're adding mask to the **transmitter** action here - for the transmitter this doesn't make sense (only receiver).

**Review by OttoWinter — COMMENTED** [review]
Ok for the idea, but I think we can implement it cleaner.

Instead of adding a new `mask` value, let's make `code` accept values like `00101001100111110101xxxx` (x stands for "any").

Do this by adding a new `validate_rc_switch_raw_code` method that is like `validate_rc_switch_code`, but also accepts `x`

File: esphome/components/remote_base/rc_switch_protocol.h (hunk 1 above)
Comment (mtl010957) [reply esphome/components/remote_base/rc_switch_protocol.h]: OOPs, thanks, that explains why it showed up in the transmitter code, I'll fix that.

**Comment by mtl010957** [comment]
I agree, your approach with the "don't care" bits seems a lot simpler for the user to configure and ujnderstand. I'll rework and let you know when it's ready for review. Thanks.

File: esphome/components/remote_base/rc_switch_protocol.h [hunk 2]
```diff
@@ -169,6 +171,7 @@ class RCSwitchRawReceiver : public RemoteReceiverBinarySensorBase {
     this->code_ = decode_binary_string(code);
     this->nbits_ = code.size();
   }
+  void set_mask(const std::string &code) { this->mask_ = decode_binary_string_mask(code); }
```
Comment (OttoWinter) [inline esphome/components/remote_base/rc_switch_protocol.h]: Move this into the `set_code` function two lines above. No need to add this as a second method when the input is the same.

Comment (mtl010957) [reply esphome/components/remote_base/rc_switch_protocol.h]: Makes sense, thanks. Done.

**Comment by mtl010957** [comment]
This is ready except there seems to be some AsyncTCP dependency issue causing the test1 to fail, I'll check back again later.

**Review by OttoWinter — APPROVED** [review]

**Pull request merged**